File: quick_sort.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class quick_sort {

    public static void main(String[] args) throws IOException {
        String inputFile = "dataset_2500000.csv";
        String outputFile;

        List<String[]> list = new ArrayList<>();

        //----Read data from CSV file----
        try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] rowParts = line.split(",", 2);  // split into 2 parts only (number, text)
                list.add(rowParts);
            }
        }

        // Set outputFile name based on actual list size
        outputFile = "quick_sort_" + list.size() + "_java.csv";

        //----Start timing the sort----
        long start = System.nanoTime();
        quickSort(list, 0, list.size() - 1);
        long end = System.nanoTime();

        //----Write sorted data to output file----
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(outputFile))) {
            for (String[] row : list) {
                bw.write(row[0] + "," + row[1]);
                bw.newLine();
            }
        }

        //----Print the runtime in seconds----
        System.out.printf("Quick Sort runtime for dataset size %d: %.3f seconds%n",
                list.size(), (end - start) / 1e9);
    }

    //-----Quicksort implementation---------------------
    static void quickSort(List<String[]> S, int left, int right) {
        if (left < right) {
            int pi = partition(S, left, right);
            quickSort(S, left, pi - 1);
            quickSort(S, pi + 1, right);
        }
    }

    //-----Partition method using String[]-----//
    static int partition(List<String[]> S, int left, int right) {
        LinkedList<String[]> L = new LinkedList<>();
        LinkedList<String[]> E = new LinkedList<>();
        LinkedList<String[]> G = new LinkedList<>();

        String[] pivot = S.get(right);
        int pivot_num = Integer.parseInt(pivot[0]);

        for (int i = left; i <= right; i++) {
            String[] e = S.get(i);
            int eval = Integer.parseInt(e[0]);

            if (eval < pivot_num) {
                L.add(e);
            } else if (eval == pivot_num) {
                E.add(e);
            } else {
                G.add(e);
            }
        }

        int index = left;
        for (String[] e : L) {
            S.set(index++, e);
        }
        for (String[] e : E) {
            S.set(index++, e);
        }
        for (String[] e : G) {
            S.set(index++, e);
        }

        return left + L.size(); // Return the index where pivot section starts
    }
}

File: quick_sort.py
import csv
import time
import sys
import os

sys.setrecursionlimit(2000000)          # Increase recursion depth for large datasets



#----read the entire CSV file into a list of Data objects----
def read_csv(filename):
    data_list = []
    with open(filename, 'r') as csvfile:
        reader = csv.reader(csvfile)
        for row in reader:
            data_list.append([int(row[0]), row[1]])  # Add to the list as a Data object
    return data_list


#----Function to write the sorted Data list into a new CSV file----
def write_csv(filename, data_list):
    with open(filename, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerows(data_list)  # Write each row as [number, text]

#-----Partition method ----
def partition(arr, left, right):
    pivot = arr[right]
    pivot_num = pivot[0]

    L, E, G = [], [], []

    for i in range(left, right + 1):
        current = arr[i]
        if current[0] < pivot_num:
            L.append(current)
        elif current[0] == pivot_num:
            E.append(current)
        else:
            G.append(current)

    index = left
    for item in L:
        arr[index] = item
        index += 1
    for item in E:
        arr[index] = item
        index += 1
    for item in G:
        arr[index] = item
        index += 1

    return left + len(L)  # Return the pivot starting index


#-----QuickSort recursive implementation -----
def quick_sort(arr, left, right):
   if left < right:  # Only sort if at least two elements
       pi = partition(arr, left, right)     # Partition the list
       quick_sort(arr, left, pi - 1)        # sort left half
       quick_sort(arr, pi + 1, right)       # sort right half

#-----Main function to execute the full quick sort process-----#
def main():
    input_file = "dataset_2500000.csv"  # Input file

    # Read data from CSV file
    data_list = read_csv(input_file)

    # Set output filename using actual data size
    output_file = f"quick_sort_{len(data_list)}_python.csv"

    # Start timer (exclude I/O)
    start_time = time.perf_counter_ns()
    quick_sort(data_list, 0, len(data_list) - 1)
    # quick_sort_iterative(data_list)
    end_time = time.perf_counter_ns()

    # Write sorted data to output file
    write_csv(output_file, data_list)

    # Print runtime in seconds
    runtime_sec = (end_time - start_time) / 1e9
    print(f"Quick Sort runtime for dataset size {len(data_list)}: {runtime_sec:.3f} seconds")

#----Run main() if this file is executed directly----#
if __name__ == "__main__":
    main()
